include/boost/thread/detail/thread.hpp

```diff
@@ -487,29 +487,18 @@ namespace thread_detail {
         template <class Rep, class Period>
         bool try_join_for(const chrono::duration<Rep, Period>& rel_time)
         {
-          return try_join_until(chrono::steady_clock::now() + rel_time);
+          return try_join_until(thread_detail::internal_clock_t::now() + rel_time);
         }
 #endif
 
         template <class Clock, class Duration>
         bool try_join_until(const chrono::time_point<Clock, Duration>& t)
         {
           using namespace chrono;
-          bool joined= false;
-          do {
-            thread_detail::internal_clock_t::time_point     s_now = thread_detail::internal_clock_t::now();
-            typename Clock::duration   d = ceil<nanoseconds>(t-Clock::now());
-            if (d <= Clock::duration::zero()) return false; // in case the Clock::time_point t is already reached
-            joined = try_join_until(s_now + d);
-          } while (! joined);
-          return true;
-        }
-        template <class Duration>
-        bool try_join_until(const chrono::time_point<thread_detail::internal_clock_t, Duration>& t)
-        {
-          using namespace chrono;
-          typedef time_point<thread_detail::internal_clock_t, nanoseconds> nano_sys_tmpt;
-          return try_join_until(nano_sys_tmpt(ceil<nanoseconds>(t.time_since_epoch())));
+          thread_detail::internal_clock_t::time_point     s_now = thread_detail::internal_clock_t::now();
```

Comment:
I don't agree here. You are converting from Clock to internal_clock.

Reply:
I guess I'm not sure where the issue is. The code is taking the difference between t and Clock::now(), adding that difference to internal_clock_t::now(), and then passing the result to try_join_until(). Later on in this file, try_join_until() converts the internal clock time to a timespec and passes it to the pthreads implementation of do_try_join_until(), which expects an internal clock time.

Reply:
Oh, I see. I suspect that we need to iterate until no_timeout or Clock::now() >= tp, isn't it?

Reply:
The reason I took out and avoided code that loops until Clock::now() >= tp is because, if Clock::now() is the system clock and the system clock jumps, we could end up timing out at the wrong time.

My thought is that, even if someone passes us a system_clock time, if we have access to steady_clock (because BOOST_THREAD_HAS_CONDATTR_SET_CLOCK_MONOTONIC is enabled), we should convert from system_clock to steady_clock and then use steady_clock under the covers.

The following doesn't provide a 100% reliable conversion from a system_clock time to a steady_clock time, but I think it's better than not converting to steady_clock and then sleeping or timing out at the wrong time due to a system clock jump. This is the conversion that I use throughout the code.

steady_clock_time = ( system_clock_time - system_clock::now() ) + steady_clock::now()

Reply:
I agree with the conversion. But I don't agree with the result of the function. The result should depend on whether the system timepoint has been reached or not.